fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java

```diff
@@ -255,19 +260,79 @@ private LuceneQueryClause getQueryForAndOr(@Nonnull LucenePlanState state, @Nonn
             if (childMask != null) {
                 childMask.setSatisfied(true);
             }
-            childClauses.add(childClause);
+            if (childClause instanceof LuceneBooleanQuery && ((LuceneBooleanQuery)childClause).getOccur() == occur) {
+                childClauses.addAll(((LuceneBooleanQuery)childClause).getChildren());
+                if (childClause instanceof LuceneNotQuery) {
+                    negatedChildren.addAll(((LuceneNotQuery)childClause).getNegatedChildren());
+                }
+            } else {
+                childClauses.add(childClause);
+            }
         }
         if (filterMask != null && filterMask.getUnsatisfiedFilters().isEmpty()) {
             filterMask.setSatisfied(true);
         }
+        if (!negatedChildren.isEmpty()) {
+            return new LuceneNotQuery(childClauses, negatedChildren);
```

Comment:
If someone did something like `lucene_clause_a OR NOT lucene_clause_b`, could this transformation lead to this being translated into something like `(lucene_clause_a) -(lucene_clause_b)` (effectively turning the `OR` to an `AND`)?

Reply:
I do not think so. `LuceneNotQuery` always has occur = `MUST`, so it is not possible for `negatedChildren` to be added to in the `SHOULD` case. Note how flattening only happens when occur matches.